Automattic/wp-calypso
# Shopping Cart: Convert lib/cart-values to TypeScript

#### Changes proposed in this Pull Request

This converts `client/lib/cart-values` to TypeScript. Nearly all of this diff is types only, but there were some modifications to the code necessary, which I'll enumerate here:

- `domainRegistration()` used to add the `is_domain_registration: true` property but that property is not part of `RequestCartProduct` and as far as I can tell is not used by the shopping-cart endpoint for requests (rather, it is a read-only property set by the endpoint on the `ResponseCartProduct`); this PR removes it.
- `siteRedirect()` used to allow creating a product object without a `domain`; now that function throws an error if one is not provided as a property on its `properties` argument.
- `googleApps()` used to allow creating a product object without a domain; now that function throws an error if one is not provided by either the `meta` or `domain` properties of its `properties` argument.
- `titanMailMonthly()` used to allow creating a product object without a domain; now that function throws an error if one is not provided by either the `meta` or `domain` properties of its `properties` argument.
- `shouldBundleDomainWithPlan()` now verifies that the `suggestionOrCartItem` has a product slug before calling `isDomainMapping()`.
- `isDomainMappingFree()` had a bug where it did not correctly check for a blogger plan (`isBloggerPlan()` requires a string and it was passing an object); this is now fixed.
- `getDomainPriceRule()` now verifies that the `suggestion` has a product slug before calling `isDomainMapping()`.
- `fillInSingleCartItemAttributes()` used to return `undefined` if the requested product slug was not found in the list of products; now it throws an error if that happens.

#### Testing instructions

No testing should be necessary since most of the changes are types. The changes most likely to cause problems are the errors thrown when calling `titanMailMonthly`, `googleApps`, or `siteRedirect` without a domain; therefore to review this PR one should make sure that all call sites of those functions always provide a domain argument (see the notes above for what each expects to receive).

To review the code changes, it's a little hard with the github diff since it's not aware that the files have just been renamed. However, you can view the _actual_ diff on the command line with the following commands, if the branch is checked out locally:

```
git diff 369ca87b87:client/lib/cart-values/cart-items.js client/lib/cart-values/cart-items.ts
git diff 369ca87b87:client/lib/cart-values/index.js client/lib/cart-values/index.ts
```
